### DISCUSSION TITLE
Adding feature flag to queries best practice

### DISCUSSION THREAD

**Discussion by rmeshksar** [graphql-dotnet/graphql-dotnet#3700] [Q&A]
Hi,
I want to add a feature flag to the API end-points.

I can see different approaches that could be used and wondering which one is recommended and why.

- Creating an extension for Field like RequireFeature and adding a meta data to the field and then creating a FieldMiddleware to check that meta data and reject the request if feature flag is off.
- Same as above, setting a meta data but registering a validator and add it to the ValidationRules of the execution options.
- Adding a directive called "feature" and apply if to the field and creating and registering a node visitor to check it and act accordingly.

Thanks.

**Comment by Shane32** [comment]
Generally you’ll create a FieldBuilder extension method for code-first and/or an attribute that inherits from GraphQLAttribute for type-first. Both of those options can add a directive to the field or simply add metadata.  For schema-first you’d probably use a directive. If you want to print the schema with the feature noted, you’d probably use a directive. Using metadata will be a little easier to implement and may execute slightly faster. 

During execution, the choice of FieldMiddleware or a validation rule (implemented as a node visitor) depends on what the feature does. Middleware currently has easier access to variables and state in general, and can operate before, after or instead of the default resolver. Whereas validation rule runs before any resolvers execute and in so doing will prevent the entire query from running if it fails. Also, field middleware runs for each execution of a resolver (or doesn’t run if the resolver doesn’t run), whereas validation rules run once for each field that could execute. 

For example, if you want to add a logging feature, which logs each execution of a certain field, you’ll need FieldMiddleware. But if you want to validate permissions, a validation rule is a better option as it is more optimized and ensures that the query is not run if any fields are disallowed for the current user.

**Reply by rmeshksar** [reply]
Thanks for prompt reply.
I would like to expand upon my previous question.

We do have an API which is shared and used by different clients. Each client has its own database and the API end-point can be enabled/disabled via feature flag. 
It means at run time, upon the execution of the API, the first thing that should be checked it if that feature flag for the certain end-point for that client is enabled or not, and if not an error should be returned.

If I understood your response correctly, in this case a validation rule works better. Correct?

I think we cannot use Node Visitor approach as the logic should be checked in every single call and Node Visitor only runs once. Am I understanding it correctly?

**Comment by Shane32** [comment]
If I understand correctly, it’s basically a permission based on the customer/database ID (versus the User ID). I’m guessing the user id defines the customer id or in some other way you determine which customer id is active.

I would implement this as a validation rule. I’d probably have the validation rule pull the allowed features from the customer id. Then when validating the fields it can just check against the pulled feature list. 

Actually there’s another way you could implement this. Before GraphQL executes you could write some asp.net middleware to pull the feature list from your database and then modify the ClaimsPrincipal to append claims (e.g. roles) based on the enabled features. Then GraphQL and any other ASP.NET code can use typical authorization code to match on the enabled features.

**Comment by Shane32** [comment]
The node visitor / validation rule approach runs once per field per GraphQL execution. So as long as each GraphQL request only applies to a single customer/database, then a validation rule is the recommended approach.  Whereas if variables or field arguments may change the feature list allowed for the call, then it would need to be field middleware. 

Note that a node visitor is typical method of implementing a validation rule. (For our discussion purposes here, they are the same.)
